docs/design/simplified_federation_deployment.md

```diff
@@ -0,0 +1,388 @@
+`kubeadm` simplifies Kubernetes control plane deployment. Federation
+control plane deployment builds on top of this tool. Cluster
+registration and deregistration will be facilitated through
+[`kubectl`](http://kubernetes.io/docs/user-guide/kubectl-overview/).
+
+
+#### Federation Control Plane Deployment Experience
+
+1. User provisions their cloud/compute resources: a VM or a physical
+   machine.
+2. Downloads the
+   [`kubeadm`](http://kubernetes.io/docs/getting-started-guides/kubeadm/)
+   executable on to each of those machines, virtual or physical.
+3. Runs:
+
+    ```shell
+    $ kubeadm init federation
+    ```
+
+   to set up a Kubernetes control plane and a federation control plane
+   on that machine.
+
+
+#### Cluster Registration Experience
+
+1. User downloads the
+   [`kubectl`](http://kubernetes.io/docs/getting-started-guides/kubeadm/)
+   executable on their local computer, i.e. the computer they use for
+   their day-to-day activities.
+2. Runs:
+
+    ```shell
+    $ kubectl register mycluster
+    ```
+
+   to register their cluster named `mycluster`, where `mycluster` is the
+   name of the cluster's context in the local kubeconfig.
+
+#### Cluster Deregistration Experience
+
+1. To deregister a cluster, user runs:
+
+    ```shell
+    $ kubectl deregister mycluster
+    ```
+
+### Points for Discussion:
+
+1. Is `unregister` a better verb than `deregister`?
+2. Does the verb pair `join/unjoin` make more sense in this context?
+
+
+## Design
+
+
+#### Federation Control Plane Bootstrap
+
+`kubeadm init` initializes a Kubernetes control plane. In order to do
+that, it generates the manifests for the various control plane
+components and starts them. It also generates the certificates and the
+credentials required for those components to interact with the API
+server.
+
+Federation control plane deployment builds on top of this functionality.
+Running `kubeadm init federation` first performs all the operations
+involved in bootstrapping a Kubernetes control plane, as running
+`kubeadm init` would. We henceforth call this Kubernetes cluster for
+which the control plane was initialized, the "bootstrap cluster". Upon
+initializing the bootstrap cluster, we perform the following operations
+to bootstrap a federation control plane.
+
+
+**1. Create a namespace for federation system components**
+
+* Call this namespace `federation-system` to keep it consistent with
+  Kubernetes.
+* The namespace is created in the bootstrap cluster.
+
+**2. Expose a network endpoint for the federation API server**
+
+* Create a "Loadbalancer" type Kubernetes service in the
+  `federation-system` namespace in the bootstrap cluster to expose the
+  yet to be created federation API server.
+* Wait until a load balancer IP address is allocated to the service.
+
+**3. Generate TLS certificates and credentials**
+
+* Using `kubeadm`'s PKI infrastructure, generate TLS certificates for a
+  new CA, yet to be created federation API server and federation
```

Comment:
I thought about it a bit but didn't have a strong inclination towards one or the other. If we tie it to the CA of one of the clusters, it kind of becomes confusing in HA case - "which cluster's CA is responsible for this server certificate?". It is not a big deal though, we already have a notion of a bootstrap cluster, so we can use that as a convention. Generating a new CA seemed marginally better to keep the concerns separated.